### PR TITLE
Restrict extra props

### PR DESCRIPTION
**Pull request by matthias-pichler-warrify**
Signed-off-by: Matthias Pichler <m.pichler@warrify.com>

**Please specify parts of this PR update:**

- [ ] Specification
- [x] Schema
- [x] Examples
- [ ] Extensions
- [ ] Use Cases
- [ ] Community
- [ ] CTK
- [ ] Other

**Discussion or Issue link**:

**What this PR does**:

I added a lot of `unevaluatedProperties: false` to restrict extra props

**Additional information:**

### CODE REVIEW

File: schema/workflow.yaml [hunk 1]
```diff
@@ -5,6 +5,7 @@ type: object
 properties:
   document:
     type: object
+    unevaluatedProperties: false
```
Comment (JBBianchi) [inline schema/workflow.yaml:8]: I would assume `document` could be extended. Some runtimes could use it to store metadata for instance. (e.g. https://cloud-native.slack.com/archives/CMBAZQSEP/p1721730364697849 )

Comment (cdavernas) [reply schema/workflow.yaml:8]: @JBBianchi Spot on, totally agree with you!

Comment (JBBianchi) [reply schema/workflow.yaml:8]: (I deleted my comment to make a more generic one. For the record, I was stating that `document` could be extended by a runtime to store metadata for instance.)

**Comment by JBBianchi** [comment]
I wonder if this PR isn't a bit too drastic (but it might not be). 

An important feature of the spec is its extensibility. By restricting everything, we might lose a some of the flexibility.

For instance (as I commented earlier before deletion), the `document` could be extended by some runtime with metadata (e.g.: author, tenant, group, ...).

The author of a workflow also have access to very powerful variables, namely $workflow and $task. Therefore, he can define custom properties (e.g. `$task.with.params.$filter` or `$workflow.constants.someKey`) that could be used/remapped in some custom functions.

I'm not sure how to verify we didn't restrict the author too much here. At a glance, I think it's quite ok but it's not a very empirical approach ^^'

**Comment by matthias-pichler-warrify** [comment]
> I wonder if this PR isn't a bit too drastic (but it might not be).
> 
> An important feature of the spec is its extensibility. By restricting everything, we might lose a some of the flexibility.
> 
> For instance (as I commented earlier before deletion), the `document` could be extended by some runtime with metadata (e.g.: author, tenant, group, ...).
> 
> The author of a workflow also have access to very powerful variables, namely $workflow and $task. Therefore, he can define custom properties (e.g. `$task.with.params.$filter` or `$workflow.constants.someKey`) that could be used/remapped in some custom functions.
> 
> I'm not sure how to verify we didn't restrict the author too much here. At a glance, I think it's quite ok but it's not a very empirical approach ^^'

Very good points here, however I feel like I have to counter with the fact that I came across the issue because some example workflows behaved differently than expected 😅 After updating the schema I had to fix 3 examples in the repo and none of us caught these errors. So on behalf of authors I would advocate for better feedback from tooling which need a schema. 

Maybe we could introduce a dedicated field for metadata similar to [JSON API](https://jsonapi.org/format/1.0/#document-meta) 🤔

**Comment by cdavernas** [comment]
I agree with @JBBianchi that we should preserve extensibility, without however, as pointed out by @matthias-pichler-warrify, loosing (schema-based) validation features.

> Maybe we could introduce a dedicated field for metadata similar to [JSON API](https://jsonapi.org/format/1.0/#document-meta) 🤔

Maybe we could reintroduce that property in specific places, indeed. Maybe even everywhere? I don't really have a strong opinion on that subject, so I'll leave it up to you guys 😉

**Comment by matthias-pichler-warrify** [comment]
> Maybe we could reintroduce that property in specific places, indeed. Maybe even everywhere? I don't really have a strong opinion on that subject, so I'll leave it up to you guys 😉

@cdavernas what I also thought of just now:
alternatively we could enforce a prefix for custom/meta fields? Something like `$`, `@` or `_`?

```yaml
document: {}
do:
  - getPet:
       call: http
       with: {}
       $foo: bar
```

no preference for any given prefix

**Comment by cdavernas** [comment]
> alternatively we could enforce a prefix for custom/meta fields? 

Why not? Good idea, even though I personally am not a huge fan of it, as I think it would "dirty" the definition. But again, no strong opinion here anyways.

@JBBianchi @fjtirado @ricardozanini What fo you guys think?

**Comment by fjtirado** [comment]
I think we might add a "custom" key in those places where we think implementors might add custom stuff. 

For example

```
document: {}
do:
  - getPet:
       call: http
       with: {}
       custom: 
         alternativeBaseURI: www.alternative.com
```

**Comment by matthias-pichler-warrify** [comment]
@fjtirado 
> I think we might add a "custom" key in those places where we think implementors might add custom stuff.

Yeah so `custom`, `meta` or similar. (name is still open I think). However this would be for authors not implementors

**Comment by fjtirado** [comment]
@matthias-pichler-warrify Sure authors would use them and implementors might implement it ;)

**Review by cdavernas — APPROVED** [review]
LGTM, cheers! ❤️ 

We will however need to open a new issue/PR addressing @JBBianchi's comments ;)

**Comment by JBBianchi** [comment]
I think the `callTask` shouldn't be restricted, otherwise it clashes with [custom functions](https://github.com/serverlessworkflow/specification/blob/main/dsl.md#creating-a-custom-function), doesn't it ?

**Comment by ricardozanini** [comment]
@JBBianchi is this portion of the schema are you referring too?

```yaml

- title: CallFunction
        $ref: '#/$defs/taskBase'
        type: object
        unevaluatedProperties: false
        required: [ call ]
        properties:
          call:
            type: string
            not:
              enum: ["asyncapi", "grpc", "http", "openapi"]
            description: The name of the function to call.
          with:
            type: object
            additionalProperties: true
            description: A name/value mapping of the parameters, if any, to call the function with.
```

If so, I think it's correct. It's the `with` property that requires any object and the schema seems to allow it.

A way of validating this PR is to add an example that breaks the rule on purpose. We should always add examples/test cases to PRs that change the schema. I'll try to reinforce that in the CI.

**Comment by JBBianchi** [comment]
> @JBBianchi is this portion of the schema are you referring too?
> 
> ```yaml
> - title: CallFunction
>         $ref: '#/$defs/taskBase'
>         type: object
>         unevaluatedProperties: false
>         required: [ call ]
>         properties:
>           call:
>             type: string
>             not:
>               enum: ["asyncapi", "grpc", "http", "openapi"]
>             description: The name of the function to call.
>           with:
>             type: object
>             additionalProperties: true
>             description: A name/value mapping of the parameters, if any, to call the function with.
> ```
> 
> If so, I think it's correct. It's the `with` property that requires any object and the schema seems to allow it.


You're absolutely right.

**Review by ricardozanini — APPROVED** [review]

**Comment by matthias-pichler-warrify** [comment]
Is there anything else needed to merge this?

**Comment by cdavernas** [comment]
@matthias-pichler-warrify Not that I know of. Merged, thanks for the reminder!

**Pull request merged**